cli/pkg/deviceauth/authenticator.go

```diff
@@ -0,0 +1,246 @@
+}
+
+// DeviceCodeResponse encapsulates the response for obtaining a device code.
+type DeviceCodeResponse struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationURI         string `json:"verification_uri"`
+	VerificationCompleteURI string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	PollingInterval         int    `json:"interval"`
+}
+
+// DeviceVerification represents the response from verifying a device.
+type DeviceVerification struct {
+	DeviceCode              string
+	UserCode                string
+	VerificationURL         string
+	VerificationCompleteURL string
+	CheckInterval           time.Duration
+	ExpiresAt               time.Time
+}
+
+// DeviceAuthenticator performs the authentication flow for logging in.
+type DeviceAuthenticator struct {
+	client   *http.Client
+	BaseURL  *url.URL
+	Clock    clock.Clock
+	ClientID string
+}
+
+// New returns an instance of the DeviceAuthenticator
+func New(authURL string) (*DeviceAuthenticator, error) {
+	baseURL, err := url.Parse(authURL)
+	if err != nil {
+		return nil, err
+	}
+
+	authenticator := &DeviceAuthenticator{
+		client:   http.DefaultClient,
+		BaseURL:  baseURL,
+		Clock:    clock.New(),
+		ClientID: OAuthClientID,
+	}
+
+	return authenticator, nil
+}
+
+// VerifyDevice performs the device verification API calls.
+func (d *DeviceAuthenticator) VerifyDevice(ctx context.Context) (*DeviceVerification, error) {
+	req, err := d.newFormRequest(ctx, "oauth/device_authorization", url.Values{
+		"client_id": []string{d.ClientID},
+		"scope":     []string{"full_account"},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	res, err := d.client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Body.Close()
+
+	if _, err = checkErrorResponse(res); err != nil {
+		return nil, err
+	}
+
+	deviceCodeRes := &DeviceCodeResponse{}
+	err = json.NewDecoder(res.Body).Decode(deviceCodeRes)
+	if err != nil {
+		return nil, errors.Wrap(err, "error decoding device code response")
+	}
+
+	checkInterval := time.Duration(deviceCodeRes.PollingInterval) * time.Second
+	if checkInterval == 0 {
+		checkInterval = time.Duration(5) * time.Second
+	}
+
+	expiresAt := d.Clock.Now().Add(time.Duration(deviceCodeRes.ExpiresIn) * time.Second)
+
+	return &DeviceVerification{
+		DeviceCode:              deviceCodeRes.DeviceCode,
+		UserCode:                deviceCodeRes.UserCode,
+		VerificationCompleteURL: deviceCodeRes.VerificationCompleteURI,
+		VerificationURL:         deviceCodeRes.VerificationURI,
+		ExpiresAt:               expiresAt,
+		CheckInterval:           checkInterval,
+	}, nil
+}
+
+// GetAccessTokenForDevice uses the device verification response to fetch an
+// access token.
+func (d *DeviceAuthenticator) GetAccessTokenForDevice(ctx context.Context, v DeviceVerification) (*OAuthTokenResponse, error) {
```

Comment:
If `VerifyDevice` returns a `*DeviceVerification`, this should probably have `v *DeviceVerification` instead